In [1]:
def my_cool_function(arg1, arg2):
    arg1 += 'bar'
    arg2 += 'buzz'
    return arg1 + arg2

my_cool_function('foo', 'fizz')

'foobarfizzbuzz'

In [51]:

def init(N, f):
    pop = []
    frac = N*f
    print(frac)
    for i in range(N):
        if i < frac:
            pop.append(1)
        else:
            pop.append(0)
    return(pop)


In [39]:
init(10, 0.5)

5.0


[1, 1, 1, 1, 1, 0, 0, 0, 0, 0]

In [6]:
import random

In [52]:
def step(pop):
    N = len(pop)
    new_pop = random.choices(pop, k = N)
    return(new_pop)

In [47]:
pop = [0, 0, 0, 0, 0, 1, 1, 1, 1, 1]
step(pop)

[1, 0, 0, 0, 0, 0, 0, 0, 1, 1]

In [52]:
pop = []

In [58]:
def wf(N, f, ngens):
    pop = init(N, f)
    for i in range(ngens):
        pop = step(pop)
    x = (sum(pop)/N)
    print(f"generations: {ngens} ; freq(a): {x}")

In [57]:
wf(100, 0.2, 10)

20.0
generations: 10 ; freq(a): 0.26


In [2]:
## Implementing the WF Model as a Class
class Population:
    pass

In [4]:
p = Population()

In [5]:
#initializing class attributes with init method 
class Population:
    def __init__(self, N, f):
        self.N = N
        self.f = f
        
        
        self.pop = []
        frac = N*f    
        for i in range(N):
            if i < frac:
                self.pop.append(1)
            else:
                self.pop.append(0)
            

In [6]:
p = Population()


TypeError: __init__() missing 2 required positional arguments: 'N' and 'f'

In [7]:
#adding default values if N and f are not specified
class Population:
    def __init__(self, N=10, f=0.2):
        self.N = N
        self.f = f
        
        
        self.pop = []
        frac = N*f    
        for i in range(N):
            if i < frac:
                self.pop.append(1)
            else:
                self.pop.append(0)

                
p= Population()
print(f"N={p.N} f={p.f}")

N=10 f=0.2


In [11]:
#creating printable representations of an object with __repr__()
#provides functionality to print objects in a readable way

class Population:
    def __init__(self, N=10, f=0.2):
        self.N = N
        self.f = f
        
        
        self.pop = []
        frac = N*f    
        for i in range(N):
            if i < frac:
                self.pop.append(1)
            else:
                self.pop.append(0)
    
    def __repr__(self):
        return f"Population(N={self.N}, f={self.f})"
    
p = Population()
p

                

Population(N=10, f=0.2)

In [26]:
#Defining a class method to implement the WF step() function

class Population:
    def __init__(self, N=10, f=0.2):
        self.N = N
        self.f = f
        
        
        self.pop = []
        frac = N*f    
        for i in range(N):
            if i < frac:
                self.pop.append(1)
            else:
                self.pop.append(0)
    
    def __repr__(self):
        return f"Population(N={self.N}, f={self.f})"
 
    def step(self, ngens=1):
        for i in range(ngens):
            self.pop = random.choices(self.pop, k=len(self.pop))


[0, 1, 0, 1, 1, 1, 1, 0, 0, 1]

In [61]:
#k=len(self.pop) is redundant ; change to k=N
class Population:
    def __init__(self, N=10, f=0.2):
        self.N = N
        self.f = f
                
        self.pop = []
        frac = N*f    
        for i in range(N):
            if i < frac:
                self.pop.append(1)
            else:
                self.pop.append(0)
    
    def __repr__(self):
        return f"Population(N={self.N}, f={self.f})"
 
    def step(self, ngens=1):
        for i in range(ngens):
            self.pop = random.choices(self.pop, k= self.N)

p = Population()
print(p.pop)
p

[1, 1, 0, 0, 0, 0, 0, 0, 0, 0]


Population(N=10, f=0.2)

In [20]:
p = Population(1000000000)

In [21]:
%%timeit
p.N

41.1 ns ± 0.78 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


In [22]:
%%timeit
len(p.pop)

85.2 ns ± 1.28 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


In [4]:
## Numpy challenge

In [ ]:
import numpy as n
import random

In [31]:
class Population:
    def __init__(self, N=10, f=0.2, with_np = False):
        #class attributes
        self.N = N
        self.f = f
        self.with_np = with_np
        
        if self.with_np:
            derived_count = round(N*f)
            self.pop = np.array([0] * (N - derived_count) + [1] * derived_count)
        else:
            derived_count = round(N*f)
            self.pop = [0] * (N - derived_count) + [1] * derived_count
                
    def __repr__(self):
        return f"Population(N={self.N}, f={self.f})"
 
    def step(self, ngens=1):
        
        if self.with_np:
            self.pop = np.random.choice(self.pop, size = self.N)
        else:
            for i in range(ngens):
                self.pop = random.choices(self.pop, k= self.N)


In [1]:
pop = Population(25)
pop_np = Population(25, with_np = True)

NameError: name 'Population' is not defined

In [48]:
%%timeit
pop.step(20)

208 µs ± 6.35 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [43]:
%%timeit
pop_np.step(20)

23.1 µs ± 804 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)


## Numpy Challenge Results

*Population size 100, ngens 1000* \
pop: 
20.9 ms ± 2.49 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

pop_np: 
24 µs ± 780 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

*Population size 50, ngens 100* \
pop: 
1.02 ms ± 12 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

pop_np: 
24 µs ± 1.53 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

*Population size 25, ngens 20* \
pop: 
208 µs ± 6.35 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

pop_np: 
23 µs ± 3.5 µs per loop 
(mean ± std. dev. of 7 runs, 10000 loops each)

*Conclusion* \
For each of these comparative tests of performance between using lists and numpy arrays, the tests using numpy arrays were an order of magnitude faster than use of lists per loop. Using numpy arrays, more loops were run

